### PR TITLE
Fix itemKeyStore dissapearing for files backed by signature

### LINKED ISSUE #14396: `ItemKeyStore` can disappear leading to `FindBackgroundReferencesInFile` not working
**Repro steps**

Provide the steps required to reproduce the problem:

1. In VS, `FSharp.Compiler.Service` solution
2. Open [`Driver/CompilerDiagnostics.fs`](https://github.com/dotnet/fsharp/blob/main/src/Compiler/Driver/CompilerDiagnostics.fs)
3. Go to [`type PhasedDiagnostic`](https://github.com/dotnet/fsharp/blob/23e22eadf193cdd7e38ea4fa68c0f76d1b14ca9b/src/Compiler/Driver/CompilerDiagnostics.fs#LL342C15-L342C15)
4. Find All References (69 references are found)
5. Change the file (e.g. add an empty line somewhere)
6. **Save** 
7. Find All References on `PhasedDiagnostics` again

**Expected behavior**

We get the same results as the first time

**Actual behavior**

Only 63 references found this time, all results from `Driver/CompilerDiagnostics.fs` are missing. 

**Related information**

`BackgroundCompiler.FindReferencesInFile` returns an empty sequence, because `ItemKeyStore` is `None` in the check results for this file.

* Re-type checking the file (e.g. after making other changes) doesn't fix it.
* Changing a file _above_ the affected file in compilation order does seem to fix it. 
* This also affects rename functionality. 
* VS Version 17.5.0 Preview 2.0 [33127.37.main]

### PR DESCRIPTION
**Pull request by 0101**
Should fix #14396 

The problem seems to be that here we clear `TcInfoExtras` and replace it with an empty one.
https://github.com/dotnet/fsharp/blob/85c5cc34d4535f96ec93708b947d87163e8c1dd1/src/Compiler/Service/IncrementalBuild.fs#L1079-L1087

This change fixes that but it could potentially cause unneeded work (?). I still don't fully understand the partial type checking system...

Another way could be to somehow try to re-trigger the type check when we find the empty key store when we're actually trying to use it.

### CODE REVIEW

**Review by vzarytovskii — APPROVED** [review]

File: tests/FSharp.Compiler.ComponentTests/FSharpChecker/FindReferences.fs [hunk 1]
```diff
@@ -240,3 +240,65 @@ and mytype = MyType
         fileName, 2, 5, 11
         fileName, 5, 13, 19
     ]
+
+/// https://github.com/dotnet/fsharp/issues/14396
+[<Fact>]
+let ``itemKeyStore dissappearance`` () =
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/FSharpChecker/FindReferences.fs:246]:
```suggestion
let ``itemKeyStore disappearance`` () =
```
not the best test name, just saying

**Review by psfinaki — APPROVED** [review]

**Review by TIHan — APPROVED** [review]

**Comment by majocha** [comment]
It started to fail for me here:

https://github.com/dotnet/dotnet/blob/c96a662957b78259a527f6c5911e31411153c960/src/fsharp/tests/FSharp.Test.Utilities/ProjectGeneration.fs#L479-L482

So it looks like `ParseAndCheckProject` didn't report this error before but does in my branch.

**Comment by 0101** [comment]
This is weird. How did it work for me and in CI? The `Custom` signature is supposed to automatically get the namespace + module from the project. But I may have only made that change in a branch that hasn't been merged yet. I will have a look at this.

**Comment by majocha** [comment]
@0101 it worked because of interplay wit enablePartialTypeChecking, the error surfaces with full typecheck.

**Comment by 0101** [comment]
I see. Anyway the tests shouldn't have invalid source. I might extract the update for ProjectGeneration from https://github.com/dotnet/fsharp/pull/14554 to a separate PR and merge it.

**Pull request merged**